### PR TITLE
SCRUM-1522 Enable loading of genes, alleles, and AGMs from direct submissions

### PR DESCRIPTION
**Pull request by markquintontulloch**
Alpha will need to be restarted after merge to rebuild database constraints dropped in migration.

### CODE REVIEW

File: src/main/java/org/alliancegenome/curation_api/jobs/executors/AgmExecutor.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        return runLoad(taxonIds, agms);
+    }
+        
+    public APIResponse runLoad(String taxonId, List<AffectedGenomicModelDTO> agms) {
+        List<String> taxonIds = new ArrayList<String>();
+        taxonIds.add(taxonId);
+        return runLoad(taxonIds, agms);
+    }
+            
+    public APIResponse runLoad(List<String> taxonIds, List<AffectedGenomicModelDTO> agms) {
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/jobs/executors/AgmExecutor.java:71]: What case does this cover? Just wondering why we would have a list of taxons?

File: src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java [hunk 2]
```diff
@@ -19,10 +20,46 @@ public Gene getByIdOrCurie(String id) {
...
-        Query jpqlQuery = entityManager.createQuery("SELECT annotation.curie FROM Gene annotation WHERE annotation.taxon.curie=:taxonId");
-        jpqlQuery.setParameter("taxonId", taxonID);
+    public List<String> findAllCuriesByTaxon(String taxonId) {
+        Query jpqlQuery = entityManager.createQuery("SELECT gene.curie FROM Gene gene WHERE gene.taxon.curie=:taxonId");
+        jpqlQuery.setParameter("taxonId", taxonId);
         return (List<String>) jpqlQuery.getResultList();
     }
+    
+    @Transactional
+    public void deleteReferencingDiseaseAnnotations(List<String> geneCuries) {
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java:30]: There may be a better method than this...

File: src/main/java/org/alliancegenome/curation_api/model/entities/AlleleDiseaseAnnotation.java [hunk 3]
```diff
@@ -19,11 +28,12 @@
...
 @Schema(name = "Allele_Disease_Annotation", description = "Annotation class representing a allele disease annotation")
 @JsonTypeName("AlleleDiseaseAnnotation")
+@OnDelete(action = OnDeleteAction.CASCADE)
 public class AlleleDiseaseAnnotation extends DiseaseAnnotation {
 
     @IndexedEmbedded(includeDepth = 1)
     @IndexingDependency(reindexOnUpdate = ReindexOnUpdate.SHALLOW)
-    @ManyToOne(cascade = {})
+    @ManyToOne
     @org.hibernate.annotations.OnDelete(action = org.hibernate.annotations.OnDeleteAction.CASCADE)
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/model/entities/AlleleDiseaseAnnotation.java:37]: Ahh ok with the class level @OnDelete do we need both of these?

File: src/main/java/org/alliancegenome/curation_api/model/entities/BiologicalEntity.java [hunk 4]
```diff
@@ -26,5 +37,8 @@ public class BiologicalEntity extends CurieAuditedObject {
     @JsonView({View.FieldsOnly.class})
     private NCBITaxonTerm taxon;
     
+    //@OneToMany(mappedBy = "diseaseGeneticModifier", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
+    //private    List<DiseaseAnnotation> diseaseGeneticModifiers;
+
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/model/entities/BiologicalEntity.java]: Are these coming back? There is the exclude in the @ToString

File: src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java [hunk 5]
```diff
@@ -36,7 +59,6 @@
 @Entity
 @Data
 @EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = true)
-//@ToString(exclude = {"genomicLocations"})
```
Comment (oblodgett) [inline src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java:39]: Was this intentional?

**Review by oblodgett — COMMENTED** [review]
Whats a fairly large change and it looks good for the most part.

File: src/main/java/org/alliancegenome/curation_api/jobs/executors/AgmExecutor.java (hunk 1 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/jobs/executors/AgmExecutor.java:71]: SGD has multiple taxon IDs in their AGM submission

File: src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java (hunk 2 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java:30]: Probably, but I couldn't find a better way of cascading deletes where @ManyToMany relationships are involved

File: src/main/java/org/alliancegenome/curation_api/model/entities/BiologicalEntity.java (hunk 4 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/model/entities/BiologicalEntity.java]: No, I will remove those and the exclude

File: src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java (hunk 5 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java:39]: It was commented out so I just removed it.

Comment (oblodgett) [reply src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java:39]: Ahh ok... I think we should bring it back the idea is that every class should have it. Might need to adjust the exclude.

File: src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java (hunk 2 above)
Comment (oblodgett) [reply src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java:30]: Deleting the object should remove all the entries in each of those tables. If not then we need to look at the correct combination of CASCADE DELETE etc. And another option is to Pull the object from the database clear all the lists save it then delete it.

Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java:30]: I tried various combinations of cascade options but was getting nowhere.  I was trying to follow the logic described here: [https://thorben-janssen.com/avoid-cascadetype-delete-many-assocations/](https://thorben-janssen.com/avoid-cascadetype-delete-many-assocations/)

File: src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java (hunk 5 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/model/entities/DiseaseAnnotation.java:39]: It's back now

File: src/main/java/org/alliancegenome/curation_api/model/entities/BiologicalEntity.java (hunk 4 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/model/entities/BiologicalEntity.java]: Done

File: src/main/java/org/alliancegenome/curation_api/model/entities/AlleleDiseaseAnnotation.java (hunk 3 above)
Comment (markquintontulloch) [reply src/main/java/org/alliancegenome/curation_api/model/entities/AlleleDiseaseAnnotation.java:37]: My understanding was that the class-level annotation was for cascading the delete of diseaseannotations with gene/allele/agmdiseaseannotations, whereas this is to cascade the deletion of referenced genes.

File: src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java (hunk 2 above)
Comment (oblodgett) [reply src/main/java/org/alliancegenome/curation_api/dao/GeneDAO.java:30]: Ahh ok... that is a good tutorial.

File: src/main/java/org/alliancegenome/curation_api/model/entities/AlleleDiseaseAnnotation.java (hunk 3 above)
Comment (oblodgett) [reply src/main/java/org/alliancegenome/curation_api/model/entities/AlleleDiseaseAnnotation.java:37]: ok

**Review by oblodgett — APPROVED** [review]

**Pull request merged**